### DISCUSSION TITLE
Forcing new user onboarding for ToS acceptance and further details

### DISCUSSION THREAD

**Discussion by kripod** [nextauthjs/next-auth#778] [Help]
**Your question**

How to create a page which is mandatory to fill after signing up? Users shall not be able to proceed with until they accept a Terms of Service (ToS) text.

**What are you trying to do**

Reading the documentation, I saw there is a [`newUser` page available](https://next-auth.js.org/configuration/pages) for welcoming newly-registered users. I would like to enforce accepting a ToS text through a checkbox, while requesting further data (e.g. name) on the first visit.

However, if someone closes the window by accident, then there seems to be no way to detect if a user is new through callbacks. Is there a chance this could be solved when using a database for sessions, without custom logic for all the protected Next.js pages?

**Feedback**

*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by kripod** [comment]
The Keycloak authentication service has a concept of [required actions](https://github.com/keycloak/keycloak-documentation/blob/master/server_admin/topics/users/required-actions.adoc). This would be really useful, e.g. making some nullable fields on the `User` model mandatory and presenting a form when they aren't available.

**Comment by shawnmclean** [comment]
Has anyone been able to accomplish this sort of onboarding?

I'd like to show users an `/onboarding` page based on some property on the user object in the database.

Is the middleware route the best method for this? What would this setup look like with callbacks?

**Reply by bluemoon** [reply]
I was just trying to use middleware w/ this exact same problem. 

There's an issue with the prisma client and nextjs not playing nicely in middleware because it is compiled using the client compiler and not the node compiler (not sure why they chose to do that)

**Reply by sangosteve** [reply]
> Has anyone been able to accomplish this sort of onboarding?
> 
> I'd like to show users an `/onboarding` page based on some property on the user object in the database.
> 
> Is the middleware route the best method for this? What would this setup look like with callbacks?

Been also looking for this sort of solution.

**Reply by Gravy59** [reply]
I would like to know if anyone has done this as well.
